docs/canvas/canvas-expression-lifecycle.asciidoc

```diff
@@ -0,0 +1,241 @@
+}
+----
+
+NOTE: You may need to expand the element’s size to see the whole output.
+
+Canvas uses this output’s "data type" to map to a specific renderer and passes the entire output into it. It’s up to the image render function in this case to produce an image on the workpad’s page. In this case, the expression produces some JSON output, but expressions can also produce other, simpler data, like a string or a number. But usually more useful results tend to be JSON.
+
+If this had been the output of the table, pie chart, or plot chart from the previous sections, it would have been much more complex, but the idea is the same; call some function, pass in the expression’s output, and render something on the page. As far as the expression, and the associated interpreter, is concerned, the expression produced an output and its job is done.
+
+Canvas uses that output to render an element, but other applications can use expressions to do pretty much anything. As stated previously, expressions simply execute functions, and the functions are all written in Javascript. That means if you can do something in Javascript, you can do it with an expression. This can include, just to name a few things, sending emails or notifications, reading from or writing to a file, controlling devices with WebUSB or Web Bluetooth, consuming an external API, really just about anything. If your Javascript works in the environment where the code will run, such as in Node.js or in a browser, you can do it with an expression.
+
+[[canvas-expressions-composing-functions-with-subexpressions]]
+=== Composing Functions with Sub-Expressions
+
+You may have noticed another syntax in examples from other sections, namely expressions inside of curly brackets. These are called sub-expressions, and they can be used to provide a calculated value to another expression, instead of just a static one.
+
+A simple example of this is when you upload your own images to a Canvas workpad. That upload becomes an asset, and that asset can be retrieved using the `asset` function. Usually you’ll just do this from the UI, adding an image element to the page and uploading your image from the control in the sidebar, or picking an existing asset from there as well. In both cases, the system will consume that asset via the `asset` function, and you’ll end up with an expression similar to this:
+
+*[NOTE] The examples in this section can’t be copied and pasted directly, since the values used throughout will not exist in your workpad*
+
+[source,text]
+----
+image dataurl={asset 3cb3ec3a-84d7-48fa-8709-274ad5cc9e0b}
+----
+
+Sub-expressions are executed before the function that uses them is executed. In this case, `asset` will be run first, it will produce a value, the base64-encoded value of the image, and that value will be used as the value for the `dataurl` argument in the <<image_fn, image>> function. So at first pass, you’ll have something like this, after the asset function executes:
```

Comment:
I would change the last sentence to:

After the asset function executes, you'll get the following output: